Abort USB discovery flows on device unplug

puddly opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Integrations that utilize USB discovery currently have a nasty bug: once a discovery has been created, it can no longer be cleared without restarting Home Assistant. This is not intuitive.

Instead, `hass.config_entries.flow.async_abort` is used to abort these pending flows when the underlying USB device is unplugged.

For discovery specifically, this is fairly safe but there is one important caveat: if a config flow _expects_ the underlying device to be unplugged during discovery, we currently do not have a simple way to work around this. Some integrations (like the ZBT-1 hardware integration) deal with USB removal at runtime [by directly subscribing to USB port events](https://github.com/home-assistant/core/blob/e19d7250d5676608c5f13f49079e3a1163e299bc/homeassistant/components/homeassistant_sky_connect/__init__.py#L54-L81). This may be a workaround for integrations where USB unplugs are part of the config flow. An alternative is for the config flow _itself_ to implement an overridable `_async_usb_device_removed` method that itself calls `self.hass.config_entries.flow.async_abort`.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

puddly committed:
Remove `async_is_plugged_in`, it is unused

puddly committed:
Remove unnecessary `get_serial_by_id`

puddly committed:
Use `usb_service_info_from_device` helper function

puddly committed:
Create `async_get_matching_discovery_flows`

puddly committed:
Remove `USBDiscovery.seen`

puddly committed:
Abort existing flows when USB devices are unplugged

puddly committed:
Avoid mutable `flows` by collecting before returning

puddly committed:
Test: allow integrations to opt into `usb_abort_discovery_on_unplug`

bdraco commented on homeassistant/components/homeassistant_connect_zbt2/manifest.json:
```diff
@@ -20,5 +20,6 @@
     }
-  ]
+  ],
+  "usb_abort_discovery_on_unplug": true
```
Should this default to true ?  It seems like we would want this by default, even if its a breaking change.

puddly replied:
I threw this commit in as a proposal for a way to keep this behavior opt-in (I'm fine breaking custom components here, USB discovery is very rare to begin with). There are a few contrived situations where this would break existing functionality.

puddly committed:
Revert "Test: allow integrations to opt into `usb_abort_discovery_on_unplug`"

This reverts commit 4f7ec84cae609bc4e977dcb45368c980d86045e7.

puddly committed:
Remove invalid `test_resolve_serial_by_id` test
`/dev/serial/by-id/` lookups are done internally by `scan_serial_ports`,
there should be no opportunity for non-unique serial port paths to make
it through and thus no need for the USB integration to do it a second
time.

puddly committed:
Add unit tests

puddly commented:
Since we're no longer touching `manifest.json`, this is ready to review.

puddly commented:
We may want to optimize `ADD_REMOVE_SCAN_COOLDOWN = 5` after this PR. Since everything is triggered via Supervisor _itself_ detecting device events, it might even be safe to remove?

emontnemery reviewed (dismissed):
It's great to abort the flows, but it's not clear to me why we need a new way to facilitate that.

puddly committed:
Remove `async_get_matching_discovery_flows`

puddly committed:
Migrate to `async_progress_by_init_data_type`

puddly committed:
Extend unit test to account for multiple matches

puddly commented on homeassistant/components/usb/__init__.py:
```diff
@@ -59,7 +58,6 @@
 
 __all__ = [
     "USBCallbackMatcher",
-    "async_is_plugged_in",
```
`async_is_plugged_in` is dead code as far as I can tell, it's not used by Core or by any custom component on GitHub. I can re-add it if you want to be safe and formally deprecate it.

puddly replied:
Just to make this simpler, I've reverted the removal of `async_is_plugged_in` and added a new API to make the function work. We can deprecate it in a follow-up PR.

puddly committed:
Revert "Remove `async_is_plugged_in`, it is unused"

This reverts commit e7c685344ba4022e889ff1624a6570e6f94ef119.

puddly committed:
Migrate `async_is_plugged_in` to new API

puddly committed:
Revert "Migrate `async_is_plugged_in` to new API"

This reverts commit 7c1c3f0050cab89083af026f89014a99ab765b51.

puddly committed:
Reapply "Remove `async_is_plugged_in`, it is unused"

This reverts commit 3423f6a22d3da7d08e50332e968661a59b155e94.

MartinHjelmare commented on tests/components/usb/test_init.py:
```diff
@@ -1055,48 +985,6 @@ async def test_cancel_initial_scan_callback(
 
 
-@pytest.mark.usefixtures("force_usb_polling_watcher")
-async def test_resolve_serial_by_id(
```
Why do we remove this test?
